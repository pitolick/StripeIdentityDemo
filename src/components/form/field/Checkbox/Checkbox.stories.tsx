import type { Meta, StoryObj } from '@storybook/react'
import { Checkbox } from './Checkbox'

const meta: Meta<typeof Checkbox> = {
  title: 'Form/Field/Checkbox/Base',
  component: Checkbox,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof Checkbox>

/**
 * 基本形
 */
export const Base: Story = {
  args: {},
}

/**
 * ラベル
 */
export const Label: Story = {
  args: {
    labelContents: 'ラベル',
  },
}

/**
 * ラベルDOM
 */
export const LabelDom: Story = {
  args: {
    labelContents: <span>ラベル</span>,
  },
}

/**
 * className
 */
export const ClassName: Story = {
  args: {
    checkboxClassName: 'border-2',
    labelClassName: 'inline-flex items-center space-x-2',
    labelContents: <span>ラベル</span>,
  },
}

/**
 * デフォルトチェック
 */
export const DefaultChecked: Story = {
  args: {
    defaultChecked: true,
    labelContents: <span>ラベル</span>,
  },
}

/**
 * ラッパー：true
 */
export const Wrapper: Story = {
  args: {
    wrapper: true,
    labelContents: <span>ラベル</span>,
  },
}

/**
 * ラッパー：false
 */
export const NoWrapper: Story = {
  args: {
    wrapper: false,
    labelContents: <span>ラベル</span>,
  },
}
