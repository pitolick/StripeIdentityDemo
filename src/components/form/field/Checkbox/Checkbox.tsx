/**
 * チェックボックス非制御コンポーネント
 * @module Checkbox
 */
import React, { forwardRef } from 'react'

/**
 * チェックボックスコンポーネントのプロパティ
 * @typedef {Object} CheckboxProps
 * @property {React.InputHTMLAttributes<HTMLInputElement>} - チェックボックスタグのHTML属性
 */
export interface CheckboxProps extends React.InputHTMLAttributes<HTMLInputElement> {
  /**
   * class名：チェックボックス
   */
  checkboxClassName?: React.HTMLAttributes<HTMLInputElement>['className']
  /**
   * class名：ラベル
   */
  labelClassName?: React.HTMLAttributes<HTMLLabelElement>['className']
  /**
   * ラベルコンテンツ
   */
  labelContents?: React.ReactNode
  /**
   * ラベル（確認画面表示用）
   */
  label?: string
  /**
   * ラッパー
   */
  wrapper?: boolean
}

/**
 * チェックボックス非制御コンポーネント
 * @param {CheckboxProps} props - チェックボックスコンポーネントのプロパティ
 * @returns {JSX.Element} - チェックボックスコンポーネント
 */
export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>((props, ref) => {
  const { checkboxClassName = undefined, labelClassName = undefined, labelContents = undefined, wrapper = true, ...checkboxProps } = props
  const checkboxClassNameProps = checkboxClassName ? checkboxClassName : ''
  const labelClassNameProps = labelClassName ? labelClassName : ''

  // ラッパーコンポーネント
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    // wrapperがtrueの場合はlabelをfalseの場合はフラグメントを返す
    return wrapper ? <label className={`${labelClassNameProps} cursor-pointer`}>{children}</label> : <>{children}</>
  }

  return (
    <>
      <Wrapper>
        <input ref={ref} {...checkboxProps} type="checkbox" className={`rounded focus:ring-transparent ${checkboxClassNameProps}`} />
        {labelContents && <span className="text-sm md:text-base">{labelContents}</span>}
      </Wrapper>
    </>
  )
})

Checkbox.displayName = 'Checkbox'
