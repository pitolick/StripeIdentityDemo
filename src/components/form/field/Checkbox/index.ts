export * from './Checkbox'
export * from './CheckboxControl'
