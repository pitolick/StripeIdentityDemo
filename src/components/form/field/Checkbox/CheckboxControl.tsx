/**
 * チェックボックス制御コンポーネント（react-hook-form）
 * @module CheckboxControl
 * @todo カードスタイルのときにTabでフォーカスした場合にアウトラインがでないのを修正
 */
import React, { forwardRef } from 'react'
import { Checkbox, CheckboxProps } from './Checkbox'
import { InputBox, InputBoxProps } from '@/components/form/parts'

/**
 * チェックボックスコンポーネントのプロパティ
 * @typedef {Object} CheckboxControlProps
 * @property {React.InputHTMLAttributes<HTMLInputElement>} - チェックボックスタグのHTML属性
 */

export interface CheckboxControlProps extends React.InputHTMLAttributes<HTMLInputElement>, Omit<InputBoxProps, 'children'> {
  /**
   * リスト
   */
  lists: CheckboxProps[]
  /**
   * ラベルのタイプ
   */
  labelType?: 'default'
  /**
   * 並び方
   */
  direction?: 'row' | 'col'
  /**
   * チェックボックスの位置
   */
  checkboxPosition?: 'center' | 'top' | 'hidden'
  /**
   * チェックボックスの色
   */
  checkboxColor?: 'black' | 'red' | 'blue' | 'gray'
}

/**
 * チェックボックス制御コンポーネント（react-hook-form）
 * @param {CheckboxControlProps} props - チェックボックスコンポーネントのプロパティ
 * @returns {JSX.Element} - チェックボックスコンポーネント
 */
export const CheckboxControl = forwardRef<HTMLInputElement, CheckboxControlProps>((props, ref) => {
  const {
    lists,
    nameId,
    label,
    note,
    required = undefined,
    error,
    labelType = 'default',
    direction = 'row',
    checkboxPosition = 'center',
    checkboxColor = 'blue',
    contentWidth,
    ...checkboxControlProps
  } = props

  const checkboxList = lists.map((list, index) => {
    const { labelContents, checkboxClassName, labelClassName, ...checkboxProps } = list
    // class名
    let checkboxClassNameProps = `${checkboxClassName ? checkboxClassName : ''}`
    let labelClassNameProps = `${labelClassName ? labelClassName : ''} inline-block relative pl-5`

    // チェックボックスの位置
    switch (checkboxPosition) {
      case 'center':
        checkboxClassNameProps = checkboxClassNameProps + ' absolute margin-auto top-1/2 transform -translate-y-1/2'
        break
      case 'top':
        checkboxClassNameProps = checkboxClassNameProps + ' absolute margin-auto'
        // ラベルのタイプによって分岐
        switch (labelType) {
          case 'default':
            checkboxClassNameProps = checkboxClassNameProps + ' top-1'
            break
        }
        break
      case 'hidden':
        checkboxClassNameProps = 'hidden'
        break
    }

    // エラー時のclass名
    let errorLabelClass = ''
    let errorCheckboxClass = ''

    // ラベルのタイプ
    switch (labelType) {
      case 'default':
        checkboxClassNameProps = checkboxClassNameProps + ' left-0'
        errorLabelClass = error ? ' text-red' : ''
        errorCheckboxClass = error ? ' bg-red-pale text-red border-red' : ''
        break
    }

    // チェックボックスの色
    switch (checkboxColor) {
      case 'black':
        checkboxClassNameProps = checkboxClassNameProps + ' checked:text-black focus:text-black'
        break
      case 'red':
        checkboxClassNameProps = checkboxClassNameProps + ' checked:text-red focus:text-red'
        break
      case 'blue':
        checkboxClassNameProps = checkboxClassNameProps + ' checked:text-blue focus:text-blue'
        break
      case 'gray':
        checkboxClassNameProps = checkboxClassNameProps + ' checked:text-gray-700 focus:text-gray-700'
        break
    }

    // カードコンポーネント
    const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
      return <>{children}</>
    }

    return (
      <div key={`${nameId}_${index}`} className={`${direction === 'col' && 'w-full'} ${direction === 'row' && contentWidth === 'full' && 'flex-auto'}`}>
        <Wrapper>
          <Checkbox
            ref={ref}
            labelContents={labelContents}
            checkboxClassName={`${checkboxClassNameProps} ${errorCheckboxClass}`}
            labelClassName={`${labelClassNameProps} ${errorLabelClass}`}
            wrapper={true}
            {...checkboxProps}
            {...checkboxControlProps}
          />
        </Wrapper>
      </div>
    )
  })

  return (
    <InputBox nameId={nameId} label={label} note={note} required={required} error={error} contentWidth={contentWidth}>
      {/* チェックボックス */}
      <div className={`${'flex-' + direction} ${contentWidth === 'full' && 'w-full'} flex flex-wrap gap-3`}>{checkboxList}</div>
    </InputBox>
  )
})

CheckboxControl.displayName = 'CheckboxControl'
