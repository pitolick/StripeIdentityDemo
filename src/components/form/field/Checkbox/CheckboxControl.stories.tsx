import type { Meta, StoryObj } from '@storybook/react'
import { CheckboxControl } from './CheckboxControl'

const meta: Meta<typeof CheckboxControl> = {
  title: 'Form/Field/Checkbox/Control',
  component: CheckboxControl,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof CheckboxControl>

/**
 * 基本形
 */
export const Base: Story = {
  args: {
    nameId: 'base',
    lists: [
      { labelContents: 'チェックボックス1', value: '1' },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
  },
}

/**
 * 基本形：デフォルトチェック
 */
export const BaseDefaultChecked: Story = {
  args: {
    nameId: 'base-default-checked',
    lists: [
      { labelContents: 'チェックボックス1', value: '1', defaultChecked: true },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
  },
}

/**
 * 基本形：DOM
 */
export const BaseDom: Story = {
  args: {
    nameId: 'base-dom',
    lists: [
      {
        labelContents: (
          <span>
            チェック
            <br />
            ボックス1
          </span>
        ),
        value: '1',
      },
      {
        labelContents: (
          <span>
            チェック
            <br />
            ボックス2
          </span>
        ),
        value: '2',
      },
      {
        labelContents: (
          <span>
            チェック
            <br />
            ボックス3
          </span>
        ),
        value: '3',
      },
    ],
  },
}

/**
 * className
 */
export const ClassName: Story = {
  args: {
    nameId: 'class-name',
    lists: [
      {
        labelContents: 'チェックボックス1',
        value: '1',
        checkboxClassName: 'border-4',
        labelClassName: 'text-5xl',
      },
      {
        labelContents: <span className="text-4xl">チェックボックス2</span>,
        value: '2',
      },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
  },
}

/**
 * ラベル
 */
export const Label: Story = {
  args: {
    nameId: 'label',
    lists: [
      { labelContents: 'チェックボックス1', value: '1' },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
    label: 'ラベル',
  },
}

/**
 * バリデーション：必須
 */
export const ValidationRequired: Story = {
  args: {
    nameId: 'validation-required',
    lists: [
      { labelContents: 'チェックボックス1', value: '1' },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
    label: 'ラベル',
    required: true,
  },
}

/**
 * バリデーション：任意
 */
export const ValidationOptional: Story = {
  args: {
    nameId: 'validation-optional',
    lists: [
      { labelContents: 'チェックボックス1', value: '1' },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
    label: 'ラベル',
    required: false,
  },
}

/**
 * バリデーション：エラー
 */
export const ValidationError: Story = {
  args: {
    nameId: 'error',
    lists: [
      { labelContents: 'チェックボックス1', value: '1', defaultChecked: true },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
    label: 'ラベル',
    required: true,
    error: { type: 'required', message: '必須項目です' },
  },
}

/**
 * 注釈
 */
export const Note: Story = {
  args: {
    nameId: 'note',
    lists: [
      { labelContents: 'チェックボックス1', value: '1' },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
    label: 'ラベル',
    note: { list: ['注釈'] },
  },
}

/**
 * 並び方：横
 */
export const DirectionRow: Story = {
  args: {
    nameId: 'direction-row',
    lists: [
      { labelContents: 'チェックボックス1', value: '1' },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
      { labelContents: 'チェックボックス4', value: '4' },
      { labelContents: 'チェックボックス5', value: '5' },
      { labelContents: 'チェックボックス6', value: '6' },
      { labelContents: 'チェックボックス7', value: '7' },
      { labelContents: 'チェックボックス8', value: '8' },
      { labelContents: 'チェックボックス9', value: '9' },
    ],
    label: 'ラベル',
    direction: 'row',
  },
}

/**
 * 並び方：縦
 */
export const DirectionColmun: Story = {
  args: {
    nameId: 'direction-colmun',
    lists: [
      { labelContents: 'チェックボックス1', value: '1' },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
    label: 'ラベル',
    direction: 'col',
  },
}

/**
 * チェックボックスの位置：中央
 */
export const CheckboxPosition: Story = {
  args: {
    nameId: 'checkbox-position',
    lists: [
      {
        labelContents: (
          <span>
            チェック
            <br />
            ボックス1
          </span>
        ),
        value: '1',
      },
      {
        labelContents: (
          <span>
            チェック
            <br />
            ボックス
            <br />2
          </span>
        ),
        value: '2',
      },
      { labelContents: <span>チェックボックス3</span>, value: '3' },
    ],
    label: 'ラベル',
    checkboxPosition: 'center',
  },
}

/**
 * チェックボックスの位置：上
 */
export const CheckboxPositionTop: Story = {
  args: {
    nameId: 'checkbox-position-top',
    lists: [
      {
        labelContents: (
          <span>
            チェック
            <br />
            ボックス1
          </span>
        ),
        value: '1',
      },
      {
        labelContents: (
          <span>
            チェック
            <br />
            ボックス
            <br />2
          </span>
        ),
        value: '2',
      },
      { labelContents: <span>チェックボックス3</span>, value: '3' },
    ],
    label: 'ラベル',
    checkboxPosition: 'top',
  },
}

/**
 * チェックボックスの位置：非表示
 */
export const CheckboxPositionHidden: Story = {
  args: {
    nameId: 'checkbox-position-hidden',
    lists: [
      {
        labelContents: (
          <span>
            チェック
            <br />
            ボックス1
          </span>
        ),
        value: '1',
      },
      {
        labelContents: (
          <span>
            チェック
            <br />
            ボックス
            <br />2
          </span>
        ),
        value: '2',
      },
      { labelContents: <span>チェックボックス3</span>, value: '3' },
    ],
    label: 'ラベル',
    checkboxPosition: 'hidden',
  },
}

/**
 * チェックボックスの色：黒
 */
export const CheckboxColorBlack: Story = {
  args: {
    nameId: 'checkbox-color-black',
    lists: [
      { labelContents: 'チェックボックス1', value: '1', defaultChecked: true },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
    label: 'ラベル',
    checkboxColor: 'black',
  },
}

/**
 * チェックボックスの色：赤
 */
export const CheckboxColorRed: Story = {
  args: {
    nameId: 'checkbox-color-red',
    lists: [
      { labelContents: 'チェックボックス1', value: '1', defaultChecked: true },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
    label: 'ラベル',
    checkboxColor: 'red',
  },
}

/**
 * チェックボックスの色：青
 */
export const CheckboxColorBlue: Story = {
  args: {
    nameId: 'checkbox-color-blue',
    lists: [
      { labelContents: 'チェックボックス1', value: '1', defaultChecked: true },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
    label: 'ラベル',
    checkboxColor: 'blue',
  },
}

/**
 * チェックボックスの色：グレー
 */
export const CheckboxColorGray: Story = {
  args: {
    nameId: 'checkbox-color-gray',
    lists: [
      { labelContents: 'チェックボックス1', value: '1', defaultChecked: true },
      { labelContents: 'チェックボックス2', value: '2' },
      { labelContents: 'チェックボックス3', value: '3' },
    ],
    label: 'ラベル',
    checkboxColor: 'gray',
  },
}
