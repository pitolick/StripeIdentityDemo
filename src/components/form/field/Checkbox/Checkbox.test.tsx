import React from 'react'
import { render } from '@testing-library/react'
import { Checkbox } from './Checkbox'

describe('Checkbox', () => {
  it('正しくレンダリングされること', () => {
    const { container } = render(
      <Checkbox id="my-checkbox" label="チェックボックス" labelContents="ラベルコンテンツ" checkboxClassName="custom-checkbox" labelClassName="custom-label" wrapper={true} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('デフォルト値が正しく設定されること', () => {
    const { container } = render(<Checkbox id="my-checkbox" />)
    expect(container).toMatchSnapshot()
  })
})
