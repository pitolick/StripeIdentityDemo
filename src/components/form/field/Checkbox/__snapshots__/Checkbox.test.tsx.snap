// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox デフォルト値が正しく設定されること 1`] = `
<div>
  <label
    class=" cursor-pointer"
  >
    <input
      class="rounded focus:ring-transparent "
      id="my-checkbox"
      type="checkbox"
    />
  </label>
</div>
`;

exports[`Checkbox 正しくレンダリングされること 1`] = `
<div>
  <label
    class="custom-label cursor-pointer"
  >
    <input
      class="rounded focus:ring-transparent custom-checkbox"
      id="my-checkbox"
      label="チェックボックス"
      type="checkbox"
    />
    <span
      class="text-sm md:text-base"
    >
      ラベルコンテンツ
    </span>
  </label>
</div>
`;
